File: leetcode/twoPointers/lcr018.cpp
#include <algorithm>
#include <functional> //包含std::greater 和 std::less
#include <iostream>
#include <limits.h>
#include <math.h>
#include <queue>
#include <stack>
#include <unordered_set>
#include <vector>

using namespace std;
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution {
public:
  bool isPalindrome(string s) {
    int l = 0, r = s.size() - 1;
    while (l < r) {
      while (!((tolower(s[l]) >= 'a' && tolower(s[l] <= 'z')) ||
               s[l] >= '0' && s[l] <= '9')) {
        l++;
        if (l > s.size()) {
          return true;
        }
      }
      while (!((tolower(s[r]) >= 'a' && tolower(s[r] <= 'z')) ||
               s[r] >= '0' && s[r] <= '9')) {
        r--;
        if (r < 0) {
          return true;
        }
      }
      if (tolower(s[l]) != tolower(s[r])) {
        return false;
      }
      l++, r--;
    }
    return true;
  }
};
int main() {
  // 示例二叉树
  Solution solution;
  string s = "A man, a plan, a canal: Panama";
  int result = solution.isPalindrome(s);
  std::cout << "result: " << result << std::endl;
}

File: leetcode/twoPointers/lcr018.py
from typing import List
from queue import Queue


class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = s.lower()
        left: int = 0
        right: int = len(s) - 1
        while left < right:
            while not (
                (s[left] >= "a" and s[left] <= "z")
                or (s[left] >= "0" and s[left] <= "9")
            ):
                left += 1
                if left >= len(s):
                    return True
            while not (
                (s[right] >= "a" and s[right] <= "z")
                or (s[right] >= "0" and s[right] <= "9")
            ):
                right -= 1
                if right < 0:
                    return True
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1
        return True


if __name__ == "__main__":
    solution = Solution()
    s = ".,"
    result = solution.isPalindrome(s)
    print(f"result = {result}")
